proposals/2020-07-28_keyring-cmm/proposal.md

```diff
@@ -0,0 +1,198 @@
+
+## Guide-level Explanation
+
+This change defines the [Keyring CMM](#keyring-cmm) and
+the [Master Key Provider CMM](#master-key-provider-cmm)
+as base CMM implementations that can be used "as is" or composed with other CMM.
+
+This change also defines a [Default CMM](#default-cmm) as a specific configuration of a CMM
+that the AWS Encryption SDK provides as a safe default.
+
+Customers SHOULD either use the [Default CMM](#default-cmm)
+or compose a CMM using the AWS Encryption SDK's provided CMM implementations
+(including the Keyring CMM).
+
+### Default CMM
+
+The Default CMM is a specific configuration of a
+CMM implementation provided by the AWS Encryption SDK.
+
+The specific CMM configuration describes a safe default that serves most use cases.
+
+The CMM configuration defined by the Default CMM is the [Keyring CMM](#keyring-cmm) as is
+(The Keyring CMM provides no additional options other than specifying an underlying Keyring).
+
+As the AWS Encryption SDK provides more CMM implementations that would benefit default use cases,
+we expect to update the Default CMM to a configuration which composes such CMMs together provide
+useful properties.
+Any update to the composition defined by Default CMM SHOULD be backwards compatable.
+
+To ensure that its use is correct by construction,
+the instantiation of the Default CMM MUST NOT take any options
+except for the keyring that should be composed with the Default CMM.
+
+### Keyring CMM
+
+The Keyring CMM is a base CMM implementation that
+[gets encryption materials](../../framework/cmm-interface.md#get-encryption-materials)
+and [decrypts materials](../../framework/cmm-interface.md#decrypt-materials)
+through use of a [keyring](../../framework/keyring-interface.md) configured on instantiation.
+
+Any desired CMM implementation that intends to use keyrings to ensure valid materials SHOULD
+compose with the Keyring CMM.
+
+### Master Key Provider CMM
+
+Any implementation that previously supported
+[master key providers](../../framework/master-key-provider-interface.md)
+MUST have a Master Key Provider CMM.
+Any implementation that does not support master key providers MUST NOT
+provide a Master Key Provider CMM.
+
+This CMM SHOULD NOT be used except for backwards compatability reasons.
+
+The Master Key Provider CMM is a base CMM implementation that
+[gets encryption materials](../../framework/cmm-interface.md#get-encryption-materials)
+and [decrypts materials](../../framework/cmm-interface.md#decrypt-materials)
+through use of a [master key provider](../../framework/master-key-provider-interface.md)
+configured on instantiation.
+
+Any desired CMM implementation that intends to use master key providers to ensure valid materials
+SHOULD compose with the Master Key Provider CMM.
+
+## Reference-level Explanation
+
+### Default CMM
+
+On initialization it MUST accept:
+
+- an underlying [Keyring](../../framework/keyring-interface.md)
```

Comment:
What about a master key provider?

Reply:
This one makes me sad, but you're right.

Reply:
No, we want MKP to be a different CMM.

Reply:
This would break backwards compatability for users who are already using a DefaultCMM constructor that takes in a MKP. It would force those users to update their code to get it to work without actually providing any benefit to them.

But perhaps it's worth it now in order for this interface to be strictly Keyrings?

My other concern is that if we do add additional behaviors to the default CMM, would we want those changes to only people who are using Keyrings? Even if MKPs are deprecated, I could see value in being able to update their DefaultCMM behavior in a non-breaking way.

Reply:
To clarify: We want the keyring CMM to only handle keyrings and the MKP CMM to only handle CMMs. The DefaultCMM is responsible for determining which one to use based on the input.